SeQuant/domain/mbpt/mr.hpp

```diff
@@ -0,0 +1,184 @@
+// clang-format on
+qninterval_t ncre_uocc(qns_t qns);
+
+// clang-format off
+/// @return the total number of creators in \p qns
+// clang-format on
+qninterval_t ncre(qns_t qns);
+
+// clang-format off
+/// @return the number of annihilators in \p qns acting on space \p s
+/// @pre `(s.type()==IndexSpace::Type::active_occupied || s.type()==IndexSpace::Type::active_unoccupied)&&s.qns()==IndexSpace::null_qns`
+// clang-format on
+qninterval_t nann(qns_t qns, const IndexSpace& s);
+
+// clang-format off
+/// @return the number of annihilators in \p qns acting on space \p s
+/// @pre `s==IndexSpace::Type::active_occupied || s==IndexSpace::Type::active_unoccupied`
+// clang-format on
+qninterval_t nann(qns_t qns, const IndexSpace::Type& s);
+
+// clang-format off
+/// @return the number of annihilators in \p qns acting on the occupied space
+// clang-format on
+qninterval_t nann_occ(qns_t qns);
+
+// clang-format off
+/// @return the number of annihilators in \p qns acting on the active space
+// clang-format on
+qninterval_t nann_act(qns_t qns);
+
+// clang-format off
+/// @return the number of annihilators in \p qns acting on the unoccupied space
+// clang-format on
+qninterval_t nann_uocc(qns_t qns);
+
+// clang-format off
+/// @return the total number of annihilators in \p qns
+// clang-format on
+qninterval_t nann(qns_t qns);
+
+/// combines 2 sets of quantum numbers using Wick's theorem
+qns_t combine(qns_t, qns_t);
+
+}  // namespace mr
+}  // namespace mbpt
+
+/// @param qns the quantum numbers to adjoint
+/// @return the adjoint of \p qns
+mbpt::mr::qns_t adjoint(mbpt::mr::qns_t);
+
+namespace mbpt {
+namespace mr {
+
+// clang-format off
+/// @brief makes a tensor-level fermionic many-body operator for use in single-reference methods
+
+/// A many-body operator has the following generic form:
+/// \f$ \frac{1}{P} T_{b_1 b_2 \dots b_B}^{k_1 k_2 \dots k_K} A^{b_1 b_2 \dots b_B}_{k_1 k_2 \dots k_K} \f$
+/// where \f$ \{B,K\} \f$ are number of bra/ket indices of \f$ T \f$ or, equivalently, the number of creators/annihilators
+/// of normal-ordered (w.r.t. the default, not necessarily Fermi vacuum) operator \f$ A \f$.
+/// Indices \f$ \{ b_i \} \f$ / \f$ \{ k_i \} \f$ are (active) unoccupied/occupied for (pure) excitation operators,
+/// are occupied/unoccupied for deexcitation operators; for general operators complete basis indices are assumed by default,
+/// unless overridden by user manually. \f$ P \f$ is the "normalization" factor and depends on the vacuum used to define \f$ A \f$,
+/// and indices \f$ \{ b_i \} \f$ / \f$ \{ k_i \} \f$.
+/// @note The choice of unoccupied indices/spaces can be controlled by the default Formalism:
+/// - if `get_default_formalism().sum_over_uocc() == SumOverUocc::Complete` IndexSpace::complete_unoccupied will be used instead of IndexSpace::active_unoccupied
+/// - if `get_default_formalism().csv() == CSVFormalism::CSV` will use cluster-specific (e.g., PNO) unoccupied indices
```

Comment:
Does this apply to MR operators as well? I'm not sure that the concept can be straight-forwardly generalized to the MR case (though I also haven't looked into it).